app(feat): launch drop tip wizard logic

mjhuff opened the pull request:
Closes RAUT-795

<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

This PR adds logic for wiring up the drop tip modal in additional locations. It also refactors the nice-modal-react context providers to the appropriate app/ODD locations to take advantage of top-level context. It includes some drive by refactoring as well - primarily adding missing tests.

A few notes of interest/controversy:
1) Because the OT-2 provides no way to validate that a tip has been removed, and we probably shouldn't always assume the Flex sensor is accurate for determining whether a tip is attached, the banner/ODD modal logic assumes that once a user begins the drop tip flow that the tip will be removed or the user is no longer interested with tip removal (and therefore the alert should be removed). 
2) If a user has tips on both the left and right pipettes and chooses not to skip the drop tip wizard (clicks dismiss on the dekstop banner or presses skip on the ODD), BUT the user exists a drop tip wizard flow prematurely, the banner/modal popup will still exist for the second pipette. 
3) The one optimization for getPipettesWithTipAttached is if both pipettes are used in a run, check for tip status for both pipettes on the same command loop iteration. 

### Protocol Header Entrypoint 
<img width="922" alt="Screenshot 2023-10-20 at 11 18 29 AM" src="https://github.com/Opentrons/opentrons/assets/64858653/8a532328-535a-4d69-9d8c-1e53092317b2">

### ODD InstrumentDetail Entrypoint
<img width="728" alt="Screenshot 2023-10-20 at 11 18 09 AM" src="https://github.com/Opentrons/opentrons/assets/64858653/b69e6d57-63c7-4080-9da6-2f9691648baf">

### ODD RunSummary Entrypoint

<img width="509" alt="Screenshot 2023-10-20 at 11 17 48 AM" src="https://github.com/Opentrons/opentrons/assets/64858653/fc92d8f4-7705-46a0-b45a-10d3257a5212">

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Test Plan

Protocols That May Be Useful:
[pick_up_tip.py.zip](https://github.com/Opentrons/opentrons/files/13055798/pick_up_tip.py.zip)
[Just Pipettes.py.zip](https://github.com/Opentrons/opentrons/files/13055797/Just.Pipettes.py.zip)

- Place tips on one/both of the pipettes and run the just pipettes protocol. After the run the banner should appear (Desktop) or clicking return to dashboard/run again will open the modal (ODD).
- Dismissing the banner will cause the drop tip modal not to render if it is not clicked in a few seconds. Clicking "skip" on the ODD will cause the drop tip banner on the desktop app to disappear. 
- Run the pick up tip protocol and press the estop some time AFTER the tip has attached. Verify the banner/ODD modal appears. 
- Run the pick up tip protocol and press the estop BEFORE the tip has attached (but before the pickUpTip command has been performed). Verify there is no banner/ODD modal.
<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->

# Changelog
- Added ODD overflow menu on the InstrumentDetail page.
- Wired up the drop tip wizard on the Desktop & ODD apps.
<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests
Test on a physical robot, please! AppUI is configured for these protocols.
<!--
Describe any requests for your reviewers here.
-->

# Risk assessment
low
<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

mjhuff committed:
fix(app): relocate nice-modal provider to DesktopApp and OnDeviceDisplayApp

Certain hooks within modals require ApiHostProvider, which is context provided in
OnDeviceDisplayApp. Relocate modal rendering so hooks may utilize hostname.

mjhuff committed:
refactor(app): update PipetteData type instrumentModel property as PipetteModel type

mjhuff committed:
refactor(app): rename home-gantry icon to reset-position

Rename the home-gantry icon to reset-position, which is in-line with designs. Because this icon now
launches drop-tip flows, a more semantic name is appropriate.

mjhuff committed:
feat(app): wire up drop tip wizard entry from odd instrument detail overflow menu

Adds a new overflow menu on the instrument detail page with recalibrate and drop tip options.

mjhuff committed:
refactor(app): update ODD InstrumentDetail entry point given design discussion

mjhuff committed:
feat(app): add a drop tip banner

mjhuff committed:
feat(app): wire up protocolrunheader with drop tip wizard

mjhuff committed:
feat(app): determine if pipettes have tip attached after a protocol run

mjhuff committed:
feat(app): wire up odd runsummary page drop tip wizard entry point

mjhuff committed:
refactor i18n text and state instead of refs

mjhuff committed:
refactor(app): add missing testing

mjhuff committed:
there will be no pickup tip in place

mjhuff committed:
do not delete pyyaml

mjhuff committed:
lint fix

koji commented on app/src/assets/localization/en/drop_tip_wizard.json:
nit
recommend run sort on VSCode.

mjhuff replied:
Discussed w/ Nick. His PR (which will be merged in alongside mine) will fix this!

koji commented on app/src/assets/localization/en/run_details.json:
```diff
@@ -89,33 +89,34 @@
   "run_protocol": "Run Protocol",
   "run_status": "Status: {{status}}",
   "run_time": "Run Time",
-  "run": "Run",
```
Generally this change makes sense but currently we are using sort lines asc on VSCode and it sorts like below

run_failed
run

instead of 

run
run_failed

I think we can solve this after 7.1.0 release.

mjhuff replied:
Totally agree - IDK why my VSCode sorts opposite yours, but we should enforce some sort of standard.

koji commented on app/src/atoms/MenuList/index.tsx:
```diff
@@ -21,7 +21,7 @@ export const MenuList = (props: MenuListProps): JSX.Element | null => {
     <LegacyModalShell
       borderRadius={BORDERS.borderRadiusSize4}
-      width="21.875rem"
+      width="max-content"
```
thanks!

koji commented on app/src/organisms/Devices/ProtocolRun/ProtocolDropTipBanner.tsx:
```diff
@@ -0,0 +1,58 @@
+      paddingRight={SPACING.spacing16}
+    >
+      <Flex flexDirection={DIRECTION_COLUMN}>
+        <StyledText as="p" fontWeight={FONT_WEIGHT_SEMIBOLD}>
```
we can use `TYPOGRAPHY.fontWeightSemiBold` instead of `FONT_WEIGHT_SEMIBOLD`

```suggestion
        <StyledText as="p" fontWeight={TYPOGRAPHY.fontWeightSemiBold}>
```

koji commented on app/src/organisms/Devices/ProtocolRun/ProtocolDropTipBanner.tsx:
```diff
@@ -0,0 +1,58 @@
+          </StyledText>
+          <Btn
+            textAlign={ALIGN_START}
+            fontSize={TYPOGRAPHY.fontSizeP}
```
Do we need this?
StyledText uses `as="p"`

koji commented on app/src/organisms/DropTipWizard/TipsAttachedModal.tsx:
```diff
@@ -0,0 +1,108 @@
+import { Modal } from '../../molecules/Modal'
+import { DropTipWizard } from '.'
+
+import type { ModalHeaderBaseProps } from '../../molecules/Modal/types'
+import type { PipetteData } from '@opentrons/api-client'
+import type { PipetteModelSpecs, RobotType } from '@opentrons/shared-data'
```
nit

```suggestion
import type { PipetteData } from '@opentrons/api-client'
import type { PipetteModelSpecs, RobotType } from '@opentrons/shared-data'
import type { ModalHeaderBaseProps } from '../../molecules/Modal/types'
```

koji commented on app/src/organisms/DropTipWizard/TipsAttachedModal.tsx:
```diff
@@ -0,0 +1,108 @@
+interface TipsAttachedModalProps {
+  mount: PipetteData['mount']
+  instrumentModelSpecs: PipetteModelSpecs
+  robotType: RobotType
```
Question
Do we need robotType?  From import statements, this modal is for the touchscreen app which means robotType is always "OT-3 Standard".

Will this component be used for OT-2?

mjhuff replied:
Yep! We do need it for DropTipWizard. It's a bit hard to see, but we spread in the props on L94.

koji commented on app/src/organisms/DropTipWizard/TipsAttachedModal.tsx:
```diff
@@ -0,0 +1,108 @@
+              />
+              <SmallButton
+                flex="1"
+                buttonType="primary"
```
nit
buttonType can be removed since the default is primary.

```suggestion
                
```

koji commented on app/src/organisms/DropTipWizard/TipsAttachedModal.tsx:
```diff
@@ -0,0 +1,108 @@
+          <DropTipWizard
+            {...props}
+            closeFlow={() => {
+              onCloseClick?.((pipettesWithTip: any[]) => {
```
Can we avoid using `any[]`?

koji commented on app/src/organisms/DropTipWizard/__tests__/TipsAttachedModal.test.tsx:
```diff
@@ -0,0 +1,81 @@
+
+import { i18n } from '../../../i18n'
+import { handleTipsAttachedModal } from '../TipsAttachedModal'
+import { LEFT } from '@opentrons/shared-data'
```
this line can be right after `renderWithProviders`

koji commented:
For a test of time, you may need to call a convert function in test.

mjhuff commented:
> For a test of time, you may need to call a convert function in test.

Thanks for noticing. I think I'll remove the exact time testing, since the clock functionality is tested elsewhere and I'm just testing for the existing of elements on a page 👍

mjhuff committed:
feedback and non-clock dependent testing

mjhuff commented:
Note: The PD test failure appears unrelated to this PR 👍

mjhuff committed:
refactor(robot-server): check if protocol run is in terminal state instead of played

Drop tip flows introduce the need to create a maintenence run while a protocol run has played.
Instead of checking for the played status, check if a protocol run is in a terminal state.

mjhuff commented:
@sanni-t Here are the python-sided changes we discussed in our sync on Monday.

mjhuff committed:
Merge branch 'edge' into app_launch-drop-tip-wizard-logic

b-cooper commented on app/src/pages/OnDeviceDisplay/RunSummary.tsx:
```diff
@@ -109,17 +129,37 @@ export function RunSummary(): JSX.Element {
+      handleTipsAttachedModal(
+        mount,
+        specs,
+        ROBOT_MODEL_OT3,
```
This should really be `FLEX_ROBOT_TYPE` from shared data, that ROBOT_MODEL_OT3 happens to be the same string, but it's a deprecated const

b-cooper reviewed (approved):
Changes look good 🪶

sanni-t commented on robot-server/robot_server/maintenance_runs/router/base_router.py:
```diff
@@ -153,10 +153,10 @@ async def create_run(
         check_estop: Dependency to verify the estop is in a valid state.
     """
-    if protocol_run_has_been_played:
+    if not protocol_run_is_terminal:
```
This will change the behavior so that you can only create maintenance runs once a protocol run is completed. You want to also allow starting maintenance runs when a protocol run has been created but not played.

sanni-t commented on api/src/opentrons/protocol_engine/state/commands.py:
```diff
@@ -642,6 +642,17 @@ def has_been_played(self) -> bool:
+            return True
+        else:
+            return False
+
```
You can check this same condition with just a `self._state.run_result is not None`

sanni-t commented on robot-server/robot_server/runs/dependencies.py:
```diff
@@ -121,15 +121,15 @@ async def get_engine_store(
     except AssertionError:
         return False
-    return protocol_run_state.commands.has_been_played()
+    return protocol_run_state.commands.get_is_terminal()
```
You can change this to something like:

```
async def get_is_okay_to_create_maintenance_run(
    engine_store: EngineStore = Depends(get_engine_store),
) -> bool:
    """Whether a maintenance run can be created if a protocol run already exists. """
    try:
        protocol_run_state = engine_store.engine.state_view
    except AssertionError:
        return False
    return not protocol_run_state.commands.has_been_played() or protocol_run_state.commands.get_is_terminal()
```

sanni-t reviewed (changes requested):
The robot server change in this PR will block you from creating maintenance runs before a protocol is started. That means no LPC or calibrations during protocol run setup. Added suggestions for fixing this.

mjhuff committed:
feedback pt 2

mjhuff commented:
@sanni-t I retested with these changes and things seem to be working as expected. No issue with calibration/LPC/maintenance runs now 👍 

@b-cooper Can confirm the useEffect in ProtocolRunHeader isn't firing egregious network requests 👍

mjhuff committed:
words

sanni-t reviewed (approved):
🙌 🧙‍♂️ Thanks for making the changes!
